internal/provisioner/kops_provisioner_cluster.go

```diff
@@ -376,6 +376,55 @@ func (provisioner *KopsProvisioner) UpgradeCluster(cluster *model.Cluster) error
 	}
 	defer kops.Close()
 
+	switch kopsMetadata.Version {
+	case "latest", "":
+		err = kops.UpgradeCluster(kopsMetadata.Name)
+		if err != nil {
+			return err
+		}
+	default:
+		setValue := fmt.Sprintf("spec.kubernetesVersion=%s", kopsMetadata.Version)
+		err = kops.SetCluster(kopsMetadata.Name, setValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	instanceGroups, err := kops.GetInstanceGroupsJSON(kopsMetadata.Name)
+	if err != nil {
+		return err
+	}
+	for _, ig := range instanceGroups {
+		if ig.Spec.Image != kopsMetadata.AMI {
+			if len(kopsMetadata.AMI) == 0 {
```

Comment:
This if statement & everything inside would make a good helper function, which would in turn make this parent function a bit easier to read.

Reply:
Will do